clang/labs/lab1/kosarev_egor/always_inline.cpp

```diff
@@ -0,0 +1,70 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+class AddAlwaysInlineConsumer : public clang::ASTConsumer {
+public:
+  bool hasCondition(clang::Stmt *S) {
+    if (!S) {
+      return false;
+    }
+
+    if (llvm::isa<clang::IfStmt>(S) || llvm::isa<clang::SwitchStmt>(S) ||
+        llvm::isa<clang::WhileStmt>(S) || llvm::isa<clang::DoStmt>(S) ||
+        llvm::isa<clang::ForStmt>(S)) {
+      return true;
+    }
+
+    for (clang::Stmt *Child : S->children()) {
+      if (hasCondition(Child)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool HandleTopLevelDecl(clang::DeclGroupRef D) override {
+    clang::FunctionDecl *FD = nullptr;
+
+    for (clang::Decl *Decl : D) {
+      FD = clang::dyn_cast<clang::FunctionDecl>(Decl);
+      if (FD) {
+        if (!FD->hasAttr<clang::AlwaysInlineAttr>() &&
+            !hasCondition(FD->getBody())) {
+          FD->addAttr(
+              clang::AlwaysInlineAttr::CreateImplicit(FD->getASTContext()));
+          FD->print(llvm::outs() << "__attribute__((always_inline)) ");

```

Comment:
Could you please check that attribute is present in AST tree after adding?

Reply:
I've redone the tests

Reply:
Great, thank you.
Then printing is no more required.
Could you please remove `print` calls?

Reply:
done